torch/testing/_internal/common_device_type.py

```diff
@@ -21,7 +23,7 @@
 # [WRITING TESTS]
 #
 # Write your test class as usual except:
-#   (1) Each test method should have one of four signatures:
+#   (1) Each test method should have one of following five signatures:
```

Comment:
Looking over the whole API holistically here, the current formulation of the API seems weirdly path-dependent on the order historically we added features.

In a typical `pytest.parametrize` style setup, you get an argument per what you parametrize over. You parametrize over three things, you get three parameters.

In contrast, let's look at the device-type testing:

```
#           (1a) testX(self, device)
#
#           (1b) @deviceCountAtLeast(<minimum number of devices to run test with>)
#                testX(self, devices)
#
#           (1c) @dtypes(<list of dtypes> or <list of tuples of dtypes>)
#                testX(self, device, dtype)
#
#           (1d) @deviceCountAtLeast(<minimum number of devices to run test with>)
#                @dtypes(<list of dtypes> or <list of tuples of dtypes>)
#                testX(self, devices, dtype)
#
#           (1e) @ops(<list of OpInfo instances>)
#                testX(self, device, dtype, op)
```

The inconsistencies seem to come from good intentions, but for example, why is the default parametrization over only device? (History?) Ops directly implies op-dependent parametrization on device and dtype; but what if I want to test all ops but I don't actually want to run all the tests on cuda (because maybe it's expensive?) And the most important point: are the decorators composable; can I mix and match them? (I don't see any examples of this in the PR--maybe it already works? It is hard to tell.)

When designing an API, we have a choice between making things hyperspecialized for the use case at hand (risking ossification for when we need to generalize in the future) versus making things general from first principles (risking not actually building a good tool that solves the problem at hand.) I don't want us to get high on abstraction, but I would like to see some overarching story that says (1) how to compose these decorators and (2) how to add new decorators orthogonally to existing decorators.